src/Swarm/Game/Location.hs

```diff
@@ -71,11 +90,68 @@ deriving instance FromJSON (V2 Int32)
 deriving instance FromJSONKey (V2 Int32)
 deriving instance ToJSONKey (V2 Int32)
 
-instance FromJSON Location where
-  parseJSON = fmap P . parseJSON
-
-instance ToJSON Location where
-  toJSON (P v) = toJSON v
+toHeading :: AbsoluteDir -> Heading
+toHeading = \case
+  DNorth -> north
+  DSouth -> south
+  DEast -> east
+  DWest -> west
+
+-- | The cardinal direction north = @V2 0 1@.
+north :: Heading
+north = V2 0 1
+
+-- | The cardinal direction south = @V2 0 (-1)@.
+south :: Heading
+south = V2 0 (-1)
+
+-- | The cardinal direction east = @V2 1 0@.
+east :: Heading
+east = V2 1 0
+
+-- | The cardinal direction west = @V2 (-1) 0@.
+west :: Heading
+west = V2 (-1) 0
+
+-- | The direction for viewing the current cell = @V2 0 0@.
+down :: Heading
+down = zero
+
+-- | The 'applyTurn' function gives the meaning of each 'Direction' by
+--   turning relative to the given heading or by turning to an absolute
+--   heading
+applyTurn :: Direction -> Heading -> Heading
+applyTurn d = case d of
+  DRelative e -> case e of
+    DLeft -> relative perp
+    DRight -> relative (fmap negate . perp)
+    DBack -> relative (fmap negate)
+    DDown -> relative (const down)
+    DForward -> relative id
+  DAbsolute e -> cardinal $ toHeading e
+ where
+  cardinal = const
+  relative = id
```

Comment:
What is the point of defining `relative = id`?  Why not just have `DLeft -> perp` and so on?  (Maybe this code was already like that but I didn't notice before.)

Reply:
No good reason, I think we originally used it for the absolute/relative distinction, but that has disappeared with the direction constructors.